### PR TITLE
Add `no-unbound-method` rule

### PR DESCRIPTION
**Pull request by andy-hanson**
#### PR checklist

- [X] Addresses an existing issue: #259
- [X] New feature, bugfix, or enhancement
  - [X] Includes tests
- [X] Documentation update

#### What changes did you make?

Added the `no-unbound-method` rule, which forbids the use of `x.foo` where `foo` is a method on a class or interface, but allows `y => x.foo(y)`.

### CODE REVIEW

File: src/rules/noUnboundMethodRule.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        optionsDescription: "Not configurable.",
+        options: null,
+        optionExamples: ["true"],
+        type: "functionality",
+        typescriptOnly: true,
+        requiresTypeInfo: true,
+    };
+    /* tslint:enable:object-literal-sort-keys */
+
+    public static FAILURE_STRING = "'this' may be unbound when a method is used as a property without being invoked.";
```
Comment (nchen63) [inline src/rules/noUnboundMethodRule.ts]: How about `Avoid referencing unbound class methods which may cause unintentional scoping of 'this'"

Comment (andy-hanson) [inline src/rules/noUnboundMethodRule.ts]: Removed "class" from that since it may be on an interface.

**Comment by nchen63** [comment]
nice, thanks!

**Comment by adidahiya** [comment]
Nothing like closing a 2 year old feature request 😃 

Can we add tests for JSX usage? As per https://github.com/palantir/tslint/issues/259#issuecomment-156529024 and https://github.com/palantir/tslint-react/issues/63

**Pull request merged**